From 529095e910ca568ce2d4b32a30ef7a9cf977b025 Mon Sep 17 00:00:00 2001
From: Kevin Hao <kexin.hao@windriver.com>
Date: Mon, 11 Mar 2013 17:46:01 +0800
Subject: [PATCH 11/12] powerpc/fsl_book3e: add the support to boot from thread1 in kdump boot

In kdump boot, we can only boot from the cpu which get a panic.
That means we have to disable the thread0 if the panic occurs on
thread1. We also can't migrate the kdump boot sequence to a thread0
as what we do in kexec boot.

Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 arch/powerpc/include/asm/reg_booke.h |    1 +
 arch/powerpc/kernel/exceptions-64e.S |   33 +++++++++++++++++++++++----------
 arch/powerpc/kernel/misc_64.S        |   23 ++++++++++++++++++++++-
 arch/powerpc/platforms/85xx/smp.c    |    4 ++--
 4 files changed, 48 insertions(+), 13 deletions(-)

diff --git a/arch/powerpc/include/asm/reg_booke.h b/arch/powerpc/include/asm/reg_booke.h
index 17e831c..663c90f 100644
--- a/arch/powerpc/include/asm/reg_booke.h
+++ b/arch/powerpc/include/asm/reg_booke.h
@@ -725,6 +725,7 @@
 #define SPRN_TENSR	0x1b5	/* Thread Enable Status Register */
 #define SPRN_TENS	0x1b6	/* Thread Enable Set Register */
 #define SPRN_TENC	0x1b7	/* Thread Enable Clear Register */
+#define SPRN_TIR	0x1be	/* Thread Indentification Register */
 
 #define TEN_THREAD(x)	(1 << x)
 
diff --git a/arch/powerpc/kernel/exceptions-64e.S b/arch/powerpc/kernel/exceptions-64e.S
index 2f3c19d..29477fc 100644
--- a/arch/powerpc/kernel/exceptions-64e.S
+++ b/arch/powerpc/kernel/exceptions-64e.S
@@ -1372,29 +1372,42 @@ _GLOBAL(init_core_book3e)
 
 _GLOBAL(fsl_enable_threads)
 BEGIN_FTR_SECTION
-	/* Disable the other thread */
+	/* Disable the other thread. In kdump boot, we may need to enable
+	 * thread0 here if the thread1 is the boot cpu. */
+#ifdef CONFIG_CRASH_DUMP
+	mfspr	r3,SPRN_TIR
+	cmpwi	0,r3,0
+	li	r3,1
+	bne	10f
+#endif
 	li	r3,2
-	mtspr	SPRN_TENC,r3
+10:	mtspr	SPRN_TENC,r3
 
-1:	mfspr	r3,SPRN_TENSR
-	andi.	r3,r3,2
+1:	mfspr	r4,SPRN_TENSR
+	and.	r4,r4,r3
 	bne	1b
 
 #ifndef CONFIG_PPC_DISABLE_THREADS
 	/* Configure the MSR per the default */
-	LOAD_REG_IMMEDIATE(r3, MSR_KERNEL);
-	MTTMR(TMRN_IMSR1, 3);
+	LOAD_REG_IMMEDIATE(r4, MSR_KERNEL);
 
 	/*
 	 * Set the NIA for the secondary thread to
 	 * generic_secondary_thread_init
 	 */
-	LOAD_REG_ADDR(r3, .fsl_secondary_thread_init);
-	MTTMR(TMRN_INIA1, 3);
+	LOAD_REG_ADDR(r5, .fsl_secondary_thread_init);
+
+	cmpwi	0,r3,1
+	beq	20f
+	MTTMR(TMRN_IMSR1, 4);
+	MTTMR(TMRN_INIA1, 5);
+	b	30f
+
+20:	MTTMR(TMRN_IMSR0, 4);
+	MTTMR(TMRN_INIA0, 5);
 
 	/* Release the other thread. It will spin until kick_cpu is called */
-	li	r3, 2
-	mtspr	SPRN_TENS, r3
+30: mtspr	SPRN_TENS, r3
 #endif
 END_FTR_SECTION_IFSET(CPU_FTR_SMT)
 
diff --git a/arch/powerpc/kernel/misc_64.S b/arch/powerpc/kernel/misc_64.S
index dd06392..1601897 100644
--- a/arch/powerpc/kernel/misc_64.S
+++ b/arch/powerpc/kernel/misc_64.S
@@ -470,7 +470,28 @@ _GLOBAL(disable_kernel_fp)
 _GLOBAL(kexec_wait)
 #ifdef CONFIG_PPC_BOOK3E
 BEGIN_FTR_SECTION
-	/* Disable all the secondary threads */
+	/* Disable all the secondary threads. In kdump boot, we will
+	 * disable the thread0 if there is a panic on thread1 */
+#ifdef CONFIG_KEXEC
+	LOAD_REG_ADDR(r4,crashing_cpu)
+	lwz	r4,0(r4)
+	cmpwi	0,r4,-1
+	beq	10f
+	lhz	r5,PACAPACAINDEX(r13)
+	LOAD_REG_ADDR(r6,threads_per_core)
+	lwz	r6,0(r6)
+	addi	r6,r6,-1
+	andc	r4,r4,r6
+	andc	r5,r5,r6
+	cmpw	r4,r5
+	bne	10f
+	mfspr	r4,SPRN_TIR
+	li	r5,1
+	sld	r4,r5,r4
+	mtspr	SPRN_TENC,r4
+	b	$
+10:
+#endif
 	andi.	r4,r3,1
 	beq	1f
 	li	r4,2
diff --git a/arch/powerpc/platforms/85xx/smp.c b/arch/powerpc/platforms/85xx/smp.c
index 1c7aa50..01b842b 100644
--- a/arch/powerpc/platforms/85xx/smp.c
+++ b/arch/powerpc/platforms/85xx/smp.c
@@ -577,8 +577,8 @@ static void mpc85xx_smp_machine_kexec(struct kimage *image)
 	}
 #endif
 
-	if (cpu_has_feature(CPU_FTR_SMT))
-		set_cpus_allowed_ptr(current, cpumask_of(boot_cpuid));
+	if (cpu_has_feature(CPU_FTR_SMT) && (crashing_cpu == -1))
+			set_cpus_allowed_ptr(current, cpumask_of(boot_cpuid));
 
 	default_machine_kexec(image);
 }
-- 
1.7.0

